### PR TITLE
Fix identity key storing optimistically

### PR DESCRIPTION
**Pull request by devceline**
# Changes
- Refactor persist logic to defer persistence till after key server returns `200`
- Added check for signature validity  (not empty)
- Change access of `identityKeys.identityKeys` (the actual store) to be public 
- Add dev deps to allow successful testing on ci (fixex for circular dep)
 
Resolves https://github.com/WalletConnect/walletconnect-utils/issues/151

### CODE REVIEW

File: misc/identity-keys/package.json [hunk 1]
```diff
@@ -52,6 +57,7 @@
     "@walletconnect/did-jwt": "2.0.3",
     "@walletconnect/types": "^2.10.1",
     "@walletconnect/utils": "^2.10.1",
-    "axios": "^1.3.5"
+    "axios": "^1.3.5",
+    "isomorphic-unfetch": "^4.0.2"
```
Comment (bkrem) [inline misc/identity-keys/package.json:61]: Should be devDep?

File: misc/identity-keys/test/identity-keys.test.ts [hunk 2]
```diff
@@ -1,7 +1,87 @@
 import "mocha";
+import "isomorphic-unfetch";
```
Comment (bkrem) [inline misc/identity-keys/test/identity-keys.test.ts:2]: Could probably just use axios (which is already a dep) for the `fetch` call needed in here but will leave it up to you 👍

File: misc/identity-keys/src/identity-keys.ts [hunk 3]
```diff
@@ -42,14 +42,26 @@ export class IdentityKeys implements IIdentityKeys {
...
     const publicKey = await ed25519.getPublicKey(privateKey);
 
     const pubKeyHex = ed25519.utils.bytesToHex(publicKey).toLowerCase();
     const privKeyHex = ed25519.utils.bytesToHex(privateKey).toLowerCase();
-    this.core.crypto.keychain.set(pubKeyHex, privKeyHex);
-    return [pubKeyHex, privKeyHex];
+
+    return {
+      pubKeyHex,
+      presist: async () => {
```
Comment (bkrem) [inline misc/identity-keys/src/identity-keys.ts:54]:
```suggestion
      persist: async () => {
```

File: misc/identity-keys/src/identity-keys.ts [hunk 4]
```diff
@@ -42,14 +42,26 @@ export class IdentityKeys implements IIdentityKeys {
...
+        // Deferring presistence to caller to only presist after success
```
Comment (bkrem) [inline misc/identity-keys/src/identity-keys.ts:55]:
```suggestion
        // Deferring persistence to caller to only persist after success
```

File: misc/identity-keys/src/identity-keys.ts [hunk 5]
```diff
@@ -96,30 +109,28 @@ export class IdentityKeys implements IIdentityKeys {
...
 
-        // Storing keys after signature creation to prevent having false statement
-        // Eg, onSign failing / never resolving but having identity keys stored.
-        this.identityKeys.set(accountId, {
-          identityKeyPriv: privKeyHex,
-          identityKeyPub: pubKeyHex,
-          accountId,
-        });
+        if (!signature) {
+          throw new Error("Provided an empty signature");
```
Comment (bkrem) [inline misc/identity-keys/src/identity-keys.ts:113]:
```suggestion
          throw new Error(`Provided an invalid signature. Expected a string but got: ${signature}`);
```

File: misc/identity-keys/src/identity-keys.ts [hunk 6]
```diff
@@ -96,30 +109,28 @@ export class IdentityKeys implements IIdentityKeys {
...
+              },
             },
-          },
-        });
-
-        if (response.status !== 200) {
-          throw new Error(`Failed to register on keyserver ${response.status}`);
+          });
+        } catch (e) {
+          throw new Error(`Failed to register on keyserver`);
```
Comment (bkrem) [inline misc/identity-keys/src/identity-keys.ts:129]: Should log what happened with the keyserver:

```suggestion
          throw new Error(`Failed to register on keyserver: ${e}`);
```

File: misc/identity-keys/test/identity-keys.test.ts [hunk 7]
```diff
@@ -1,7 +1,87 @@
...
+    const encodedIdentity = encodeEd25519Key(identity).split(":")[2];
+
+    const fetchUrl = `${DEFAULT_KEYSERVER_URL}/identity?publicKey=${encodedIdentity}`;
+
+    const fetchedFromKeyServer = await fetch(fetchUrl);
+
+    expect(fetchedFromKeyServer.status).eq(200);
+  });
+
+  it("does not presist identity keys that failed to register", async () => {
```
Comment (bkrem) [inline misc/identity-keys/test/identity-keys.test.ts:52]:
```suggestion
  it("does not persist identity keys that failed to register", async () => {
```

**Review by bkrem — APPROVED** [review]
LGTM generally, just some minor corrections needed 👍

Thanks for setting up a proper test suite here, much needed 💯

File: misc/identity-keys/test/identity-keys.test.ts (hunk 2 above)
Comment (devceline) [reply misc/identity-keys/test/identity-keys.test.ts]: yep, good idea

**Pull request merged**